core-ktx/src/main/java/com/amplifyframework/kotlin/Amplify.kt

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package com.amplifyframework.kotlin
```

Comment:
`.api`, `.datastore`, etc. _are_ in the core module, in the existing codebase.

For this PR, I've arranged all of the new API specifications into the same package, com.amplifyframework.kotlin`. So, for example, core uses `com.amplifyframework.api` for API contracts in core, but this package uses `com.amplifyframework.kotlin` instead.

If I do update the package names, I think I'd want to do like:
1. com.amplifyframework.core.kotlin.Amplify
2. com.amplifyframework.auth.kotlin.Auth
3. com.amplifyframework.datastore.kotlin.DataStore
4. etc.

Personally I'm not sure if its an improvement. What do you think?

Reply:
Thanks for your detailed explanation!  

This PR adds a `core-ktx` module, which provides Kotlin extensions for the `core` module.  In the future, we might want to add a `aws-datastore-ktx` module, which provides Kotlin extensions for the `aws-datastore` module.  For example, we might want let customers provide a coroutine that returns a `QueryPredicate`, when configuring sync expressions for selective sync in their DataStoreConfiguration.  Basically I just want to make sure we have a logical path to support more Kotlin extensions in the future.

Putting all Kotlin extension interfaces in the same package (`com.amplifyframework.kotlin`) seems like an anti-pattern though, for all the same reasons that putting your entire project in a single package would be a bad idea (name collisions, limiting scope, etc).

After you laid it out above though, I don't like my original suggestion to insert `.kotlin` as the 4th component of the package.  What if, instead, we insert it as the 3rd component, like this?

| `core` interface | `core-ktx` interface |
|----|-----|
| `com.amplifyframework.core.Amplify` | `com.amplifyframework.kotlin.core.Amplify` |
| `com.amplifyframework.auth.AuthCategoryBehavior` | `com.amplifyframework.kotlin.auth.Auth` |
| `com.amplifyframework.datastore.DataStoreCategoryBehavior` | `com.amplifyframework.kotlin.datastore.DataStore` |
| etc... | etc... |

Then, there is no ambiguity around what the package name for a Kotlin extension is, given the package of the corresponding Java interface.

> If a future all-Kotlin plugin is created, I doubt it would use `kotlin` in its namespace, either.

I agree.   My suggestion above only applies to these Kotlin extensions where we are providing a layer on top of the Java interfaces.

Reply:
We chatted about this offline and agree to go with the `com.amplifyframework.kotlin.<whatever_was_in_core>` approach.